File: MyBakery/app/src/main/java/com/nsoft/mybakery/CustomerFragment.java
package com.nsoft.mybakery;

import android.app.AlertDialog;
import android.database.Cursor;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.nsoft.mybakery.databinding.FragmentCustomerBinding;

import java.util.ArrayList;
import java.util.List;

public class CustomerFragment extends Fragment {
    private FragmentCustomerBinding binding;
    private CustomerRecyclerViewAdapter myAdapter;
    private DBHelper dbHelper;
    private List<Customer> customerList;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentCustomerBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        // my code here ----------------------------------------------------------------------------
        dbHelper = new DBHelper(getContext());
        customerList = new ArrayList<>();
        myAdapter = new CustomerRecyclerViewAdapter(getContext(), customerList, new CustomerRecyclerViewAdapter.ClickListener() {
            @Override
            public void onClick(int id) {

                new AlertDialog.Builder(getContext())
                        .setTitle("Confirm delete")
                        .setMessage("Do you really want to delete this customer?")
                        .setPositiveButton("Yes", (dialog, which) -> {
                            dbHelper.deleteCustomerById(id);
                            fetchCustomers();
                        })
                        .setNegativeButton("No", (dialog, which) -> {
                            dialog.dismiss();
                        })
                        .show();
            }
        }, new CustomerRecyclerViewAdapter.LongClickListener() {
            @Override
            public void onLongClick(Customer customer) {
                View updateCustomerDialogView = LayoutInflater.from(getContext()).inflate(R.layout.update_customer_dialog, null, false);

                AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
                builder.setView(updateCustomerDialogView);
                builder.setCancelable(false);

                AlertDialog dialog = builder.create();
                dialog.show();

                dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);

                ImageView updateDialogCustomerImage = updateCustomerDialogView.findViewById(R.id.updateDialogCustomerImage);
                EditText updateDialogCustomerNameEditText = updateCustomerDialogView.findViewById(R.id.updateDialogCustomerNameEditText);
                EditText updateDialogCustomerAddressEditText = updateCustomerDialogView.findViewById(R.id.updateDialogCustomerAddressEditText);
                EditText updateDialogCustomerNumberEditText = updateCustomerDialogView.findViewById(R.id.updateDialogCustomerNumberEditText);
                Button updateDialogAddButton = updateCustomerDialogView.findViewById(R.id.updateDialogAddButton);
                Button updateDialogCancelButton = updateCustomerDialogView.findViewById(R.id.updateDialogCancelButton);

                updateDialogCustomerNameEditText.setText(customer.getName());
                updateDialogCustomerAddressEditText.setText(customer.getAddress());
                updateDialogCustomerNumberEditText.setText("" + customer.getNumber());

                updateDialogCancelButton.setOnClickListener(v -> {
                    dialog.dismiss();
                });

                updateDialogAddButton.setOnClickListener(v -> {

                    String updatedName = updateDialogCustomerNameEditText.getText().toString().trim();
                    String updatedAddress = updateDialogCustomerAddressEditText.getText().toString().trim();
                    String updatedNumber = updateDialogCustomerNumberEditText.getText().toString().trim();

                    if (updatedName.isEmpty() && updatedAddress.isEmpty() && updatedNumber.isEmpty()) {
                        Toast.makeText(getContext(), "Please fill all fields...", Toast.LENGTH_SHORT).show();
                    } else {
                        dbHelper.updateCustomer("profile", updatedName, updatedAddress, updatedNumber, customer.getId());
                        dialog.dismiss();
                        fetchCustomers();
                    }
                });

            }
        });

        fetchCustomers();

        binding.customerRecyclerView.addItemDecoration(new ProductItemSpacingDecorator());
        binding.customerRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.customerRecyclerView.setAdapter(myAdapter);

        binding.addCustomer.setOnClickListener(v -> {
            View customerAddDialView = LayoutInflater.from(getContext()).inflate(R.layout.add_customer_dialog, null, false);

            AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(getContext());
            dialogBuilder.setView(customerAddDialView);
            dialogBuilder.setCancelable(false);

            AlertDialog dialog = dialogBuilder.create();
            dialog.show();

            Button cancelButton = customerAddDialView.findViewById(R.id.dialogCancelButton);
            Button addButton = customerAddDialView.findViewById(R.id.dialogAddButton);
            ImageView profileImageView = customerAddDialView.findViewById(R.id.dialogCustomerImage);
            EditText nameEditText = customerAddDialView.findViewById(R.id.dialogCustomerNameEditText);
            EditText addressEditText = customerAddDialView.findViewById(R.id.dialogCustomerAddressEditText);
            EditText numberEditText = customerAddDialView.findViewById(R.id.dialogCustomerNumberEditText);

            cancelButton.setOnClickListener(v1 -> {
                dialog.dismiss();
            });

            addButton.setOnClickListener(v1 -> {
                String customerProfile = "";
                String customerName = nameEditText.getText().toString().trim();
                String customerAddress = addressEditText.getText().toString().trim();
                String customerNumber = numberEditText.getText().toString().trim();

                if (customerName.isEmpty() && customerAddress.isEmpty() && customerNumber.isEmpty()) {
                    Toast.makeText(getContext(), "Please fill all fields...", Toast.LENGTH_SHORT).show();
                } else {
                    dbHelper.addCustomer(customerProfile, customerName, customerAddress, customerNumber);
                    dialog.dismiss();
                    fetchCustomers();
                }

            });

            dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);
        });

        binding.searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void afterTextChanged(Editable s) {

            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (s.toString().isEmpty()) {
                    fetchCustomers();
                } else {
                    getCustomerByName(s.toString());
                }
            }
        });

        // -----------------------------------------------------------------------------------------
    }


    private void fetchCustomers() {
        Cursor cursor = dbHelper.getAllCustomers();
        if (cursor != null && cursor.getCount() > 0) {
            customerList.clear();
            while (cursor.moveToNext()) {
                int id = cursor.getInt(0);
                String profile = cursor.getString(1);
                String name = cursor.getString(2);
                String address = cursor.getString(3);
                String number = cursor.getString(4);

                customerList.add(new Customer(id, profile, name, address, number));
            }
            myAdapter.notifyDataSetChanged();
        }
    }

    private void getCustomerByName(String name) {
        Cursor cursor = dbHelper.getCustomerByName(name);
        if (cursor != null && cursor.getCount() > 0) {
            customerList.clear();
            while (cursor.moveToNext()) {
                int id = cursor.getInt(0);
                String customerProfile = cursor.getString(1);
                String customerName = cursor.getString(2);
                String customerAddress = cursor.getString(3);
                String customerNumber = cursor.getString(4);

                customerList.add(new Customer(id, customerProfile, customerName, customerAddress, customerNumber));
            }
            myAdapter.notifyDataSetChanged();
        } else {
//            Toast.makeText(getContext(), "No data found!", Toast.LENGTH_SHORT).show();
        }
    }

}

File: MyBakery/app/src/main/java/com/nsoft/mybakery/fragments/CustomerFragment.kt
package com.nsoft.mybakery.fragments

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.Toast
import androidx.core.widget.doAfterTextChanged
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.nsoft.mybakery.R
import com.nsoft.mybakery.adapterdecorators.ItemSpacingDecorator
import com.nsoft.mybakery.adapters.CustomerRecyclerViewAdapter
import com.nsoft.mybakery.databinding.FragmentCustomerBinding
import com.nsoft.mybakery.models.Customer
import com.nsoft.mybakery.viewmodels.MyViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class CustomerFragment : Fragment() {
    private lateinit var binding: FragmentCustomerBinding
    private val myViewModel: MyViewModel by viewModels()
    private var toast: Toast? = null
    private var searchJob: Job? = null

    private val myAdapter by lazy {
        CustomerRecyclerViewAdapter(
            object : CustomerRecyclerViewAdapter.ClickListener {

                override fun deleteClickListener(customer: Customer) {
                    AlertDialog.Builder(context)
                        .setTitle("Confirm deletion!")
                        .setMessage("Customer details\nName: ${customer.name}\nAddress: ${customer.address}\nMobile: ${customer.number}\n\nDo you really want to delete this customer?")
                        .setPositiveButton("Yes") { dialog, _ ->
                            myViewModel.deleteCustomer(customer.id)
                            dialog.dismiss()
                        }
                        .setNegativeButton("No") { dialog, _ ->
                            dialog.dismiss()
                        }
                        .show()
                }

                override fun itemLongClick(customer: Customer) {
                    val customerUpdateDialView =
                        LayoutInflater.from(context)
                            .inflate(R.layout.add_customer_dialog, null, false)

                    val dialogBuilder = AlertDialog.Builder(context)
                    dialogBuilder.setView(customerUpdateDialView)
                    dialogBuilder.setCancelable(false)

                    val dialog = dialogBuilder.create()
                    dialog.show()
                    dialog.window?.setBackgroundDrawableResource(android.R.color.transparent)

                    val cancelButton =
                        customerUpdateDialView.findViewById<Button>(R.id.dialogCancelButton)
                    val updateButton =
                        customerUpdateDialView.findViewById<Button>(R.id.dialogAddButton)
                    val profileImageView =
                        customerUpdateDialView.findViewById<ImageView?>(R.id.dialogCustomerImage)
                    val nameEditText =
                        customerUpdateDialView.findViewById<EditText>(R.id.dialogCustomerNameEditText)
                    val addressEditText =
                        customerUpdateDialView.findViewById<EditText>(R.id.dialogCustomerAddressEditText)
                    val numberEditText =
                        customerUpdateDialView.findViewById<EditText>(R.id.dialogCustomerNumberEditText)

                    updateButton.text = "Update"
                    nameEditText.setText(customer.name)
                    addressEditText.setText(customer.address)
                    numberEditText.setText(customer.number)


                    cancelButton.setOnClickListener {
                        dialog.dismiss()
                    }

                    updateButton.setOnClickListener {
                        val customerProfile = ""
                        val customerName = nameEditText.text.toString().trim()
                        val customerAddress = addressEditText.text.toString().trim()
                        val customerNumber = numberEditText.text.toString().trim()

                        if (customerName.isEmpty() || customerAddress.isEmpty() || customerNumber.isEmpty()) {
                            showToast("Please fill all fields...")
                            return@setOnClickListener
                        }

                        val customer =
                            Customer(
                                customer.id,
                                customerProfile,
                                customerName,
                                customerAddress,
                                customerNumber
                            )
                        myViewModel.updateCustomer(customer)
                        dialog.dismiss()

                    }
                }
            }
        )
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentCustomerBinding.inflate(inflater, container, false)
        return binding.getRoot()
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        // my code here ----------------------------------------------------------------------------

        // adapters start here ---------------------------------------------------------------------------
        binding.customerRecyclerView.apply {
            layoutManager = LinearLayoutManager(requireContext())
            addItemDecoration(ItemSpacingDecorator())
            adapter = myAdapter
        }
        // adapters end here -----------------------------------------------------------------------


        // observers start here --------------------------------------------------------------------------

        myViewModel.getAllCustomers().observe(viewLifecycleOwner) { allCustomers ->
            myAdapter.setAllCustomersList(allCustomers)
            binding.totalCustomersTextView.text = "Tota: ${allCustomers.size}"
        }

        observeResult(
            myViewModel.checkCustomerInsertLiveData,
            "Customer inserted successfully.",
            "Customer insertion failed! Try again..."
        )

        observeResult(
            myViewModel.checkCustomerDeleteLiveData,
            "Customer deleted successfully.",
            "Customer deletion failed! Try again..."
        )

        observeResult(
            myViewModel.checkCustomerUpdateLiveData,
            "Customer updated successfully.",
            "Failed to update! Try again..."
        )

        // observers end here ----------------------------------------------------------------------

        // customer add starts here ----------------------------------------------------------------

        binding.addCustomer.setOnClickListener {
            val customerAddDialView =
                LayoutInflater.from(context).inflate(R.layout.add_customer_dialog, null, false)

            val dialogBuilder = AlertDialog.Builder(context)
            dialogBuilder.setView(customerAddDialView)
            dialogBuilder.setCancelable(false)

            val dialog = dialogBuilder.create()
            dialog.show()
            dialog.window?.setBackgroundDrawableResource(android.R.color.transparent)

            val cancelButton = customerAddDialView.findViewById<Button>(R.id.dialogCancelButton)
            val addButton = customerAddDialView.findViewById<Button>(R.id.dialogAddButton)
            val profileImageView =
                customerAddDialView.findViewById<ImageView?>(R.id.dialogCustomerImage)
            val nameEditText =
                customerAddDialView.findViewById<EditText>(R.id.dialogCustomerNameEditText)
            val addressEditText =
                customerAddDialView.findViewById<EditText>(R.id.dialogCustomerAddressEditText)
            val numberEditText =
                customerAddDialView.findViewById<EditText>(R.id.dialogCustomerNumberEditText)

            cancelButton.setOnClickListener {
                dialog.dismiss()
            }

            addButton.setOnClickListener {
                val customerProfile = ""
                val customerName = nameEditText.text.toString().trim()
                val customerAddress = addressEditText.text.toString().trim()
                val customerNumber = numberEditText.text.toString().trim()

                if (customerName.isEmpty() || customerAddress.isEmpty() || customerNumber.isEmpty()) {
                    showToast("Please fill all fields...")
                    return@setOnClickListener
                }

                val customer =
                    Customer(0, customerProfile, customerName, customerAddress, customerNumber)
                myViewModel.insertCustomer(customer)
                dialog.dismiss()

            }
        }

        // customer add ends here ------------------------------------------------------------------


        // search filter starts here ---------------------------------------------------------------

        binding.searchEditText.doAfterTextChanged { text ->
            searchJob?.cancel()
            searchJob = viewLifecycleOwner.lifecycleScope.launch {
                delay(100)
                myViewModel.filterCustomer(text?.trim().toString())
            }
        }

        // search filter ends here ---------------------------------------------------------------


    } // on create end here ------------------------------------------------------------------------

    fun showToast(text: String) {
        toast?.cancel()
        toast = Toast.makeText(context, text, Toast.LENGTH_SHORT)
        toast?.show()
    }

    private fun observeResult(
        liveData: MutableLiveData<Boolean?>,
        successMessage: String,
        failedMessage: String
    ) {
        liveData.observe(viewLifecycleOwner) { isSuccess ->
            if (isSuccess == null) return@observe
            if (isSuccess) {
                showToast(successMessage)
            } else {
                showToast(failedMessage)
            }
            liveData.value = null
        }
    }
}